### PR TITLE
Fix "/favicon.png" 404 bug and Customize Favicon swagger.ui(#147)

### PR DESCRIPTION
**Pull request by choipureum**
related issue: #147

This PR solved the default favicon 404 error and added custom favicon settings.

- default favicon
```html
<link rel="icon" type="image/png" href="data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAMAAABEpIrGAAAAkFBMVEUAAAAQM0QWNUYWNkYXNkYALjoWNUYYOEUXN0YaPEUPMUAUM0QVNUYWNkYWNUYWNUUWNUYVNEYWNkYWNUYWM0eF6i0XNkchR0OB5SwzZj9wyTEvXkA3az5apTZ+4C5DgDt31C9frjU5bz5uxTI/eDxzzjAmT0IsWUEeQkVltzR62S6D6CxIhzpKijpJiDpOkDl4b43lAAAAFXRSTlMAFc304QeZ/vj+ECB3xKlGilPXvS2Ka/h0AAABfklEQVR42oVT2XaCMBAdJRAi7pYJa2QHxbb//3ctSSAUPfa+THLmzj4DBvZpvyauS9b7kw3PWDkWsrD6fFQhQ9dZLfVbC5M88CWCPERr+8fLZodJ5M8QJbjbGL1H2M1fIGfEm+wJN+bGCSc6EXtNS/8FSrq2VX6YDv++XLpJ8SgDWMnwqznGo6alcTbIxB2CHKn8VFikk2mMV2lEnV+CJd9+jJlxXmMr5dW14YCqwgbFpO8FNvJxwwM4TPWPo5QalEsRMAcusXpi58/QUEWPL0AK1ThM5oQCUyXPoPINkdd922VBw4XgTV9zDGWWFrgjIQs4vwvOg6xr+6gbCTqE+DYhlMGX0CF2OknK5gQ2JrkDh/W6TOEbYDeVecKbJtyNXiCfGmW7V93J2hDus1bDfhxWbIZVYDXITA7Lo6E0Ktgg9eB4KWuR44aj7ppBVPazhQH7/M/KgWe9X1qAg8XypT6nxIMJH+T94QCsLvj29IYwZxyO9/F8vCbO9tX5/wDGjEZ7vrgFZwAAAABJRU5ErkJggg==" sizes="32x32">
```
image

- example of customfavicon
```html
<link rel="icon" type="image/png" href="https://try2explore.com/static/favicon.png" sizes="32x32">
```
image

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.Core/dist/index.html [hunk 1]
```diff
@@ -4,28 +4,24 @@
   <head>
     <meta charset="UTF-8">
     <title>[[SWAGGER_UI_TITLE]]</title>
-    <link rel="icon" type="image/png" href="./favicon-32x32.png" sizes="32x32" />
-    <link rel="icon" type="image/png" href="./favicon-16x16.png" sizes="16x16" />
+    <link rel="icon" type="image/png" href="[[SWAGGER_UI_FAVICON_URL]]" sizes="32x32" />
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.Core/dist/index.html:7]: Good approach! But we need to consider those things:

* We don't know whether the favicon's MIME type is `image/png` or `image/ico` or `image/jpg` or something else.
* We also don't know the favicon's size either.
* We also don't know how many favicons will be defined, with different sizes.

Therefore, I would suggest the entire HTML tag, `<link .../>` needs to be replaced with a placeholder `[[SWAGGER_UI_FAVICON_LINKS]]`, and change other logics in accordance with this.

**Review by justinyoo — CHANGES_REQUESTED** [review]
Thanks for your contribution! I left a major comment. Once you change that part, your other changes should be updated accordingly.

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.Core/dist/index.html (hunk 1 above)
Comment (choipureum) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.Core/dist/index.html:7]: Thank you for the fast code review! Based on the comments, I'll PR it again.😀

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.Core/Configurations/DefaultOpenApiCustomUIOptions.cs [hunk 2]
```diff
@@ -33,6 +42,23 @@ public DefaultOpenApiCustomUIOptions(Assembly assembly)
...
+        {
+            "<link rel=\"icon\" type=\"image/png\" href=\"dist.favicon-32x32.png\" sizes=\"32x32\" />",
+            "<link rel=\"icon\" type=\"image/png\" href=\"dist.favicon-16x16.png\" sizes=\"16x16\" />"
+        };
+
+        /// <inheritdoc/>
+        public virtual async Task<IEnumerable<string>> GetFaviconMetaTagsAsync()
+        {
+            var metaTags = new List<string>();
+            (this.CustomFaviconMetaTags as List<string>).ForEach(async x =>
```
[the remark this answers is not in the document]
Comment (choipureum) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.Core/Configurations/DefaultOpenApiCustomUIOptions.cs]: Thank you. I changed it to SELECT!

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.Core/Configurations/DefaultOpenApiCustomUIOptions.cs [hunk 3]
```diff
@@ -55,8 +81,37 @@ public virtual async Task<string> GetJavaScriptAsync()
             return await this.ReadFromStream(this.CustomJavaScriptPath).ConfigureAwait(false);
         }
 
-        private async Task<string> ReadFromStream(string path)
+        /// <inheritdoc/>
+        public void GetFaviconPaths(IEnumerable<string> metatags, string hrefPattern)
         {
+            var href = new List<string>();
+            (metatags as List<string>).ForEach(x => href.Add(Regex.Match(x, hrefPattern, RegexOptions.IgnoreCase | RegexOptions.Compiled, TimeSpan.FromSeconds(1)).Groups[1].Value));
```
[the remark this answers is not in the document]
Comment (choipureum) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.Core/Configurations/DefaultOpenApiCustomUIOptions.cs]: I removed the Html-related package again and introduced the regular expression. However, we changed it to use the regular expression within the resolve method.

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.Core/Resolvers/OpenApiCustomUIResolver.cs [hunk 4]
```diff
@@ -18,7 +19,7 @@ public static class OpenApiCustomUIResolver
         /// </summary>
         /// <param name="assembly">The executing assembly instance.</param>
         /// <returns>Returns the <see cref="IOpenApiCustomUIOptions"/> instance resolved.</returns>
-        public static IOpenApiCustomUIOptions Resolve(Assembly assembly)
+        public static IOpenApiCustomUIOptions Resolve(Assembly assembly, string hRefPattern)
```
[the remark this answers is not in the document]
Comment (choipureum) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.Core/Resolvers/OpenApiCustomUIResolver.cs]: It's the logic of determining 'href' within the current Resolve method, so would it be better to move it inside BuildAsnyc? 🤔

**Comment by choipureum** [comment]
There are many changes, so I will send a new pr.

**Comment by CSymes** [comment]
Heya - was that new PR ever made? I'm looking for this functionality, but doesn't look like it ever got added.

**Pull request closed without merging**